src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -868,15 +874,39 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
                 trace!("no parent found for {:?}", attr.doc);
                 (item.def_id.krate, parent_node)
             };
+
+            // In order to correctly resolve intra-doc-links we need to
+            // pick a base AST node to work from.  If the documentation for
+            // this module came from an inner comment (//!) then we anchor
+            // our name resolution *inside* the module.  If, on the other
+            // hand it was an outer comment (///) then we anchor the name
+            // resolution in the parent module on the basis that the names
+            // used are more likely to be intended to be parent names.  For
+            // this, we set base_node to None for inner comments since
+            // we've already pushed this node onto the resolution stack but
+            // for outer comments we explicitly try and resolve against the
+            // parent_node first.
+
+            // NOTE: there is an implicit assumption here that outer docs will always come
+            // before inner docs.
+            let base_node = if !seen_inner_docs && item.is_mod() && attr.style == AttrStyle::Inner {
+                // FIXME(jynelson): once `Self` handling is cleaned up I think we can get rid
+                // of `mod_ids` altogether
+                self.mod_ids.push(item.def_id);
+                seen_inner_docs = true;
+                Some(item.def_id)
+            } else {
+                parent_node
+            };
```

Comment:
This needs to also update `parent_node` or any following attributes will still use the outer scope.
```suggestion
            if !seen_inner_docs && item.is_mod() && attr.style == AttrStyle::Inner {
                // FIXME(jynelson): once `Self` handling is cleaned up I think we can get rid
                // of `mod_ids` altogether
                self.mod_ids.push(item.def_id);
                seen_inner_docs = true;
                parent_node = Some(item.def_id);
            }
```
and then change `base_node` to  `parent_node` below.